src/ethereum/london/spec.py

```diff
@@ -0,0 +1,1136 @@
+    block make logical sense with the contents of the parent block. The
+    information in the block's header must also match the corresponding
+    information in the block.
+
+    To implement Ethereum, in theory clients are only required to store the
+    most recent 255 blocks of the chain since as far as execution is
+    concerned, only those blocks are accessed. Practically, however, clients
+    should store more blocks to handle reorgs.
+
+    Parameters
+    ----------
+    chain :
+        History and current state.
+    block :
+        Block to apply to `chain`.
+    """
+    parent_header = chain.blocks[-1].header
+    validate_header(block.header, parent_header)
+    validate_ommers(block.ommers, block.header, chain)
+    (
+        gas_used,
+        transactions_root,
+        receipt_root,
+        block_logs_bloom,
+        state,
+    ) = apply_body(
+        chain.state,
+        get_last_256_block_hashes(chain),
+        block.header.coinbase,
+        block.header.number,
+        block.header.base_fee_per_gas,
+        block.header.gas_limit,
+        block.header.timestamp,
+        block.header.difficulty,
+        block.transactions,
+        block.ommers,
+        chain.chain_id,
+    )
+    ensure(gas_used == block.header.gas_used, InvalidBlock)
+    ensure(transactions_root == block.header.transactions_root, InvalidBlock)
+    ensure(state_root(state) == block.header.state_root, InvalidBlock)
+    ensure(receipt_root == block.header.receipt_root, InvalidBlock)
+    ensure(block_logs_bloom == block.header.bloom, InvalidBlock)
+
+    chain.blocks.append(block)
+    if len(chain.blocks) > 255:
+        # Real clients have to store more blocks to deal with reorgs, but the
+        # protocol only requires the last 255
+        chain.blocks = chain.blocks[-255:]
+
+
+def validate_header(header: Header, parent_header: Header) -> None:
+    """
+    Verifies a block header.
+
+    In order to consider a block's header valid, the logic for the
+    quantities in the header should match the logic for the block itself.
+    For example the header timestamp should be greater than the block's parent
+    timestamp because the block was created *after* the parent block.
+    Additionally, the block's number should be directly folowing the parent
+    block's number since it is the next block in the sequence.
+
+    Parameters
+    ----------
+    header :
+        Header to check for correctness.
+    parent_header :
+        Parent Header of the header to check for correctness
+    """
+    if header.number == MAINNET_FORK_BLOCK:
+        parent_gas_limit = parent_header.gas_limit * ELASTICITY_MULTIPLIER
+        parent_gas_target = parent_header.gas_limit
+    else:
+        parent_gas_limit = parent_header.gas_limit
+        parent_gas_target = parent_header.gas_limit // ELASTICITY_MULTIPLIER
+
+    parent_base_fee_per_gas = parent_header.base_fee_per_gas
```

Comment:
This line fails for the transition block, because `parent_header` does not have `base_fee_per_gas`.

Reply:
Updated